File: 221-Maximal Square.py
# dp
# Runtime: 88 ms, faster than 74.62% of Python3 online submissions for Maximal Square.
# Memory Usage: 13.2 MB, less than 62.79% of Python3 online submissions for Maximal Square.
class Solution:
    def maximalSquare(self, matrix: 'List[List[str]]') -> 'int':
        if not matrix: return 0
        row_len, col_len = len(matrix), len(matrix[0])
        dp = [[0 if matrix[row][col] == '0' else 1 for col in range(col_len)] for row in range(row_len)]
        for row in range(1, row_len):
            for col in range(1, col_len):
                if matrix[row][col] == '1': dp[row][col] = min(dp[row - 1][col], dp[row][col - 1], dp[row - 1][col - 1]) + 1
                else: dp[row][col] = 0
        result = max(max(row) for row in dp)
        return result ** 2


'''
dp矩阵更新后存储
    以这个点为正方形右下角时，正方形最大的满足条件的边长
'''

File: Algorithms/C++/221-Maximal_Square.cpp
// Runtime: 20 ms, faster than 76.78% of C++ online submissions for Maximal Square.
// Memory Usage: 11.2 MB, less than 48.15% of C++ online submissions for Maximal Square.
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        if (matrix.empty())
            return 0;
        int res = 0;
        vector<vector<int>> dp(matrix.size(), vector<int> (matrix[0].size(), 0));
        for (int row = 0; row < matrix.size(); ++row)
            for (int col = 0; col < matrix[0].size(); ++col) {
                if (!row || !col || matrix[row][col] == '0')
                    dp[row][col] = matrix[row][col] - '0';
                else
                    dp[row][col] = min(dp[row - 1][col], min(dp[row][col - 1], dp[row - 1][col - 1])) + 1;
                res = max(res, dp[row][col]);
            }
        return res * res;
    }
};
